metrics/metrics.go

```diff
@@ -58,3 +63,55 @@ func (m *LabelMap) GetFloat(key string) *expvar.Float {
 func CurrentFDs() int {
 	return currentFDs()
 }
+
+type histogram struct {
+	// buckets is a list of bucket boundaries, in increasing order.
+	buckets []float64
+	// buckStrings is a list of the same buckets, but as strings.
+	// To avoid allocations, this is only populated on first use.
+	bucketStrings []string
+
+	lm    *LabelMap
+	sum   *expvar.Float
+	count *expvar.Int
+}
+
+// NewHistogram returns a new histogram that reports to the given
+// expvar map under the given name.
+//
+// The buckets are the boundaries of the histogram buckets, in
+// increasing order. The last bucket is +Inf.
+func NewHistogram(name string, buckets []float64) *histogram {
+	if !slices.IsSorted(buckets) {
+		panic("buckets must be sorted")
+	}
+	lm := &LabelMap{
+		Label: "le",
+	}
+	expvar.Publish(name+"_bucket", lm)
+	return &histogram{
+		buckets:       buckets,
+		bucketStrings: make([]string, len(buckets)),
+		lm:            lm,
+		count:         expvar.NewInt(name + "_count"),
+		sum:           expvar.NewFloat(name + "_sum"),
+	}
+}
+
+// Observe records a new observation in the histogram.
+func (h *histogram) Observe(v float64) {
+	h.sum.Add(v)
+	h.count.Add(1)
+	for i, b := range h.buckets {
+		if v <= b {
+			str := h.bucketStrings[i]
+			if str == "" {
+				str = fmt.Sprintf("%.2f", b)
+				h.bucketStrings[i] = str
```

Comment:
I think this might crash if multiple goroutines `Observe` concurrently for the first time and attempt concurrent writes to the map. Might be safer to populate `bucketStrings` in `NewHistogram`?

Reply:
Done.